cnf-certification-test/faredge/suite.go

```diff
@@ -0,0 +1,95 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package faredge
+
+import (
+	"fmt"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/sirupsen/logrus"
+	"github.com/test-network-function/cnf-certification-test/cnf-certification-test/common"
+	"github.com/test-network-function/cnf-certification-test/cnf-certification-test/identifiers"
+	"github.com/test-network-function/cnf-certification-test/cnf-certification-test/results"
+	"github.com/test-network-function/cnf-certification-test/internal/clientsholder"
+	"github.com/test-network-function/cnf-certification-test/internal/crclient"
+	"github.com/test-network-function/cnf-certification-test/pkg/provider"
+	"github.com/test-network-function/cnf-certification-test/pkg/scheduling"
+	"github.com/test-network-function/cnf-certification-test/pkg/testhelper"
+)
+
+// All actual test code belongs below here.  Utilities belong above.
+var _ = ginkgo.Describe(common.FarEdgeTestKey, func() {
+	logrus.Debugf("Entering %s suite", common.FarEdgeTestKey)
+	var env provider.TestEnvironment
+	ginkgo.BeforeEach(func() {
+		env = provider.GetTestEnvironment()
+	})
+	ginkgo.ReportAfterEach(results.RecordResult)
+
+	testID, tags := identifiers.GetGinkgoTestIDAndLabels(identifiers.TestSchedOtherInSharedCpuPool)
+	ginkgo.It(testID, ginkgo.Label(tags...), func() {
+		nonGuranteedPods := env.GetNonGuaranteedPods()
+		testhelper.SkipIfEmptyAll(ginkgo.Skip, nonGuranteedPods)
+		testSchedOtherPolicyInSharedCPUPool(&env, filterNonGuaranteedContainers(nonGuranteedPods))
+	})
+})
+
+func filterNonGuaranteedContainers(pods []*provider.Pod) []*provider.Container {
+	var filteredContainers []*provider.Container
+	for _, pod := range pods {
+		filteredContainers = append(filteredContainers, pod.Containers...)
+	}
+	return filteredContainers
+}
+
+func testSchedOtherPolicyInSharedCPUPool(env *provider.TestEnvironment, containers []*provider.Container) {
+	/*
+		Get the container ID
+		Run the command chrt -p <<container-process-id>>
+		Parse the output
+	*/
+
+	for _, container := range containers {
+		nodeName := container.NodeName
+		debugPod := env.DebugPods[nodeName]
+		if debugPod == nil {
+			ginkgo.Fail(fmt.Sprintf("Debug pod not found on Node: %s", nodeName))
+		}
+		ocpContext := clientsholder.NewContext(debugPod.Namespace, debugPod.Name, debugPod.Spec.Containers[0].Name)
+		pid, err := crclient.GetPidFromContainer(container, ocpContext)
+		if err != nil {
+			logrus.Errorf("unable to get container process id due to: %v", err)
+			continue
+		}
+		logrus.Infof("The obtained process id is %d", pid)
+
+		command := fmt.Sprintf("chrt -p %d", pid)
+		stdout, stderr, err := crclient.ExecCommandContainerNSEnter(command, container)
```

Comment:
I do not believe this is always correct. pid is just one of the pid in the container pid namespace, we need to test all the pids in the pid namespace:
to get the pid namespace (container pid is 418087 in this example), 4026533163 is the pid namespace with 2 processes (4026533163 pid       **2**):
```
lsns -p 418087 -t pid  -n
4026533163 pid       2 418087 100999 /usr/share/elasticsearch/jdk/bin/java -Des.networkaddress.cache.ttl=60 -Des.networkaddress.cache.negative.ttl=10 -XX:+AlwaysPreTouch -Xss1m -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djna.nosys=tr

```
to list all the pids in the pid namespace containing this pid, run: 
```
ps -e  -o pidns,pid,args   
```
then search for the "pid namespace". 4026533163 is the namespace, with 2 pids :418087 and 418194. Each can have a different priority and scheduling:
```
4026533163  418087 /usr/share/elasticsearch/jdk/bin/java -Des.networkaddress.cache.ttl=60 -Des.networkaddress.cache.negative.ttl=10 -XX:+AlwaysPreTouch -Xss1m -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djna.nosys=true -XX:-OmitStackTraceInFastThrow -Dio.netty.noUnsafe=true -Dio.netty.noKeySetOptimization=true -Dio.netty.recycler.maxCapacityPerThread=0 -Dio.netty.allocator.numDirectArenas=0 -Dlog4j.shutdownHookEnabled=false -Dlog4j2.disable.jmx=true -Djava.locale.providers=COMPAT -Xms1g -Xmx1g -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=75 -XX:+UseCMSInitiatingOccupancyOnly -Djava.io.tmpdir=/tmp/elasticsearch-13358563585845435259 -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=data -XX:ErrorFile=logs/hs_err_pid%p.log -Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m -Des.cgroups.hierarchy.override=/ -XX:MaxDirectMemorySize=536870912 -Des.path.home=/usr/share/elasticsearch -Des.path.conf=/usr/share/elasticsearch/config -Des.distribution.flavor=default -Des.distribution.type=docker -Des.bundled_jdk=true -cp /usr/share/elasticsearch/lib/* org.elasticsearch.bootstrap.Elasticsearch -Ediscovery.type=single-node
4026533163  418194 /usr/share/elasticsearch/modules/x-pack-ml/platform/linux-x86_64/bin/controller
         -  418514 [kworker/4:1-events]
```
so for each of the pids in the pid namespace we should run chrt -p 418087 and chrt -p 418194:
```
deliedit@redhatwork ptp-operator]$ chrt -p 418194
pid 418194's current scheduling policy: SCHED_OTHER
pid 418194's current scheduling priority: 0
[deliedit@redhatwork ptp-operator]$ chrt -p 418087
pid 418087's current scheduling policy: SCHED_FIFO
pid 418087's current scheduling priority: 1

```

Reply:
Thanks! I interpreted the requirement differently as checked only for container process id. Any workload run from that container needs to be changed. I would add the suggested part shortly. :)